### DISCUSSION TITLE
Deploy with hypercorn and trio on available port, then get port inside app.

### DISCUSSION THREAD

**Discussion by elstanto** [fastapi/fastapi#11331] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def run():
    trio.run(serve, app, trio_config)
```

### Description

Hi, I'm writing a modular framework where there is a main API to which "modules" (other APIs, e.g. FastAPIs) can register and they will be mounted as redirects in sub-routes. So, as an example FastAPI "module", in lifespan() during startup it should fire off a POST request to the main API with it's name and details. This should include the address and port which FastAPI is listening on, so that it can assign the redirect correctly. Because I would like to start several modules in subprocesses, I would like to let the OS issue the next available port rather than managing the free ports myself (by specifying port 0 in Config()). In which case, how could I access the port that hypercorn was given, inside the app please? I should add that I would like the app to be cross-platform so want to try and avoid os calls.

If I had access to the socket instance I could call sock.getsockname() to get the port.

Another option is to open a socket on port 0, get the port and close it, before starting the app, and hoping nothing else takes the port in the meantime. Presumably if I loop this a few times with some exception handling then that would further reduce the risk.

Any opinions? Thanks!

### Operating System

Windows

### Operating System Details

Developing on Windows but should be cross-platform

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

3.12.2

**Comment by elstanto** [comment, marked as answer]
So I have a working version with the second option - checking for a free port then picking it:
```
def register_module(name:str, module: str, version: str, osb_address: str):
    if osb_address:
        try:
            requests.post(osb_address + "/api/box/modules", params={
                "name": name, 
                "module":module, 
                "version": version,
                "address": args.address + ":" + args.port
            })
        except HTTPException:
            raise

@asynccontextmanager
async def lifespan(app: FastAPI):
    logger.debug("Starting {} v{}".format(osb_module_name, __version__))
    register_module(args.name, osb_module_name, __version__, args.osb)
    yield
    logger.debug("Stopping {}".format(osb_module_name))

def get_free_port():
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
        s.bind(('', 0))
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        return str(s.getsockname()[1])

def run():
    if not args.port:
        args.port = get_free_port()
        print(args.port)
    trio_config.bind = [args.address + ":" + args.port]
    trio.run(serve, app, trio_config)  # type: ignore
```

I still wonder if there is a way to get the port from inside the app though. Thanks.

**Reply by YuriiMotov** [reply]
> I still wonder if there is a way to get the port from inside the app though. 

No, there is no such possibility in ASGI standard..
